### PR TITLE
secrets: pass through aws secrets reader configuration to clusterd

### PR DESCRIPTION
**Pull request by guswynn**
Fixes the problem discussed here: https://materializeinc.slack.com/archives/C04DJT084KA/p1690881663259479

### Motivation

  * This PR fixes a previously unreported bug.

### Tips for reviewer
The commits are logical and can be looked at individually

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by benesch** [comment]
Just skimming, but moving `EnvironmentId` around seems like an abstraction violation! I won't have time to dig in for another hour or two, but my hunch is that we should plumb the AWS values down more specifically, rather than moving `EnvironmentId` into `mz_repr`.

**Comment by guswynn** [comment]
@benesch the aws secrets controller already prefixes secrets with the env id, so I dont think we can avoid this abstraction break: https://github.com/MaterializeInc/materialize/blob/main/src/aws-secrets-controller/src/lib.rs#L126

**Review by alex-hunt-materialize — COMMENTED** [review]
This looks like it will work, but it's very weird in some places to be passing readers to things labeled controllers, and to use the readers to do write operations.

File: src/aws-secrets-controller/src/lib.rs [hunk 1]
```diff
@@ -184,9 +170,10 @@ impl AwsSecretsController {
 impl SecretsController for AwsSecretsController {
     async fn ensure(&self, id: GlobalId, contents: &[u8]) -> Result<(), anyhow::Error> {
         match self
+            .reader
```
[the remark this answers is not in the document]
Comment (guswynn) [reply src/aws-secrets-controller/src/lib.rs]: adjusted the naming here in the top commit

**Comment by benesch** [comment]
> @benesch the aws secrets controller already prefixes secrets with the env id, so I dont think we can avoid this abstraction break

I think that `clusterd` should take something like `aws_secret_prefix: String` as an argument, rather than independently reconstructing the secret prefix from the environment ID. That will avoid the abstraction break by treating the secret prefix as an opaque string.

**Comment by alex-hunt-materialize** [comment]
> I think that clusterd should take something like secrets_reader_aws_prefix: String as an argument, rather than independently reconstructing the secret prefix from the environment ID. That will avoid the abstraction break by treating the secret prefix as an opaque string.

This seems very reasonable. We would also need to pass in the region, for constructing the client.

**Comment by guswynn** [comment]
sounds good, ill switch it to that

**Comment by guswynn** [comment]
@benesch @alex-hunt-materialize changed to passing in things as strings, let me know what you think

File: src/environmentd/src/bin/environmentd/main.rs [hunk 2]
```diff
@@ -407,7 +408,7 @@ pub struct Args {
         ]),
         default_value("kubernetes"), // This shouldn't be possible, but it makes clap happy.
     )]
-    secrets_controller: SecretsControllerKind,
+    secrets_controller: SecretsReaderKind,
```
Comment (alex-hunt-materialize) [inline src/environmentd/src/bin/environmentd/main.rs:411]: Nitpick, naming is weird.

File: src/service/src/secrets.rs [hunk 3]
```diff
@@ -58,6 +73,59 @@ impl SecretsReaderCliArgs {
...
+                        self.secrets_reader_local_file_dir
+                            .as_ref()
+                            .expect("initialized correctly")
+                            .display()
+                    ),
+                ]
+            }
+            SecretsReaderKind::Kubernetes => {
+                vec![
+                    "--secrets-reader=local-file".to_string(),
```
Comment (alex-hunt-materialize) [inline src/service/src/secrets.rs:103]: Shouldn't this be `kubernetes` not `local-file`?

File: src/service/src/secrets.rs [hunk 4]
```diff
@@ -58,6 +73,59 @@ impl SecretsReaderCliArgs {
...
+                    "--secrets-reader=aws-secrets-manager".to_string(),
+                    format!(
+                        "--secrets-reader-aws-region={}",
+                        self.secrets_reader_aws_region
+                            .as_ref()
+                            .expect("initialized correctly")
+                    ),
+                    format!(
+                        "--secrets-reader-aws-prefix={}",
+                        self.secrets_reader_aws_region
```
Comment (alex-hunt-materialize) [inline src/service/src/secrets.rs:123]: Shouldn't this be `self.secrets_reader_aws_prefix`?

File: src/environmentd/src/bin/environmentd/main.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (guswynn) [reply src/environmentd/src/bin/environmentd/main.rs:411]: yeah i can fix that!

**Review by alex-hunt-materialize — APPROVED** [review]

File: src/service/src/secrets.rs [hunk 5]
```diff
@@ -11,53 +11,121 @@ use std::path::PathBuf;
...
+        env = "SECRETS_READER_AWS_PREFIX"
+    )]
+    pub secrets_reader_aws_prefix: Option<String>,
 }
 
-#[derive(ArgEnum, Debug, Clone)]
-enum SecretsReaderKind {
-    Process,
+#[derive(ArgEnum, Debug, Clone, Copy)]
+pub enum SecretsControllerKind {
```
Comment (benesch) [inline src/service/src/secrets.rs:56]: This rename doesn’t jive with the name of the other arguments declared here! Fine for now to get the release unblocked though.

File: src/controller/src/clusters.rs [hunk 6]
```diff
@@ -478,6 +478,7 @@ where
             ClusterRole::User => "user",
         };
         let persist_pubsub_url = self.persist_pubsub_url.clone();
+        let secrets_args = self.secrets_args.to_flags();
```
Comment (benesch) [inline src/controller/src/clusters.rs:481]: This was intentionally part of the orchestrator/service interface and not the controller/cluster interface. Fine for now to unblock the release, but I’d ultimately like to put the orchestrator back in charge of passing this parameter.

**Review by benesch — APPROVED** [review]
LGTM to unblock the release. Have a few comments about preserving original API boundaries, but we can address after the release!

File: src/controller/src/clusters.rs (hunk 6 above)
Comment (guswynn) [reply src/controller/src/clusters.rs:481]: sounds good! these args can be passed into the individual orchestrator's probably!

**Pull request merged**